test/services/ProcessCsvServiceSpec.scala

```diff
@@ -0,0 +1,390 @@
+      val boolList = Await.result(Future.sequence(result), Duration.Inf)
+      assert(boolList.forall(_.isLeft))
+      boolList.head.left.get
+        .getMessage mustBe "The file that you chose doesn’t contain any data.<br/><br/>You won’t be able to upload CSOP_OptionsGranted_V3.csv as part of your annual return."
+    }
+
+    "return a throwable when an error occurs during the file processing" in {
+      val callback = UpscanCsvFilesCallbackList(List(UpscanCsvFilesCallback(UploadId("1"), UploadedSuccessfully("CSOP_OptionsGranted_V3.csv", "no", Some(1)))))
+      val data = "2015-09-23,250,123.12,12.1234,12.1234,no,yes,AB12345678,no\n2015-09-23,250,123.12,12.1234,12.1234,no,yes,AB12345678,no\n2015-09-23,250,123.12,12.1234,12.1234,no,yes,AB12345678,no"
+      when(mockDataGenerator.setValidatorCsv(any())(any(), any(), any())).thenReturn(Left(ERSFileProcessingException(
+        "ers.exceptions.dataParser.configFailure",
+        Messages("ers.exceptions.dataParser.validatorError"),
+        optionalParams = Seq("test")
+      )))
+
+      val resultFuture = testProcessCsvService.processFiles(Some(callback), "csop", returnStubSource(_, data))
+
+      val result = Await.result(resultFuture, Duration.Inf)
+      val boolList = Await.result(Future.sequence(result), Duration.Inf)
+      assert(boolList.forall(_.isLeft))
+      boolList.head.left.get
+        .getMessage mustBe "ers.exceptions.dataParser.configFailure"
+    }
+
+  }
+
+  "extractEntityData" should {
+
+    "extract the data from an HttpResponse" in {
+      val response = HttpResponse(entity = "Test response body")
+      val resultFuture = testProcessCsvService.extractEntityData(response).runWith(Sink.head)
+
+      val result = Await.result(resultFuture, Duration.Inf)
+      result.utf8String mustBe "Test response body"
+    }
+
+    "return a failed source with an UpstreamErrorResponse if response status is not Ok (200)" in {
+      val response = HttpResponse(status = StatusCodes.InternalServerError)
+      val resultFuture = testProcessCsvService.extractEntityData(response).runWith(Sink.head)
+
+      assert(resultFuture.failed.futureValue.getMessage.contains("Illegal response from Upscan"))
+    }
+  }
+
+  "extractBodyOfRequest" should {
+
+    "extract the body from an Http response stream and read as csv" in {
+      val source = Source
+        .single(HttpResponse(entity = "0, 1, 2, 3\n4, 5"))
+
+      val resultFuture = testProcessCsvService.extractBodyOfRequest(source).runWith(Sink.seq)
+      val result = Await.result(resultFuture, Duration.Inf)
+
+      result.head match {
+        case Right(data) =>
+          data mustBe List(ByteString("0"), ByteString(" 1"), ByteString(" 2"), ByteString(" 3"))
+        case Left(ex) =>
+          throw ex
+      }
+      result(1) match {
+        case Right(data) =>
+          data mustBe List(ByteString("4"), ByteString(" 5"))
+        case Left(ex) =>
+          throw ex
+      }
```

Comment:
This looks odd to me, why is there a pattern match on the response of a test? The output should be known based on the data being fed in right?

Reply:
yep, that's just a result(index).right.get with extra steps - will replace it with that. I think it's one of those things we did when our brains were in a state of ers overdose